File: MATTHEW/leetcode/AIcoding/4.GPT-4o/cpp_solutions/87-scramble-string.cpp
class Solution {
public:
    bool isScramble(string s1, string s2) {
        int n = s1.size();
        if (n != s2.size()) return false;
        if (s1 == s2) return true;

        vector<int> count(26, 0);
        for (int i = 0; i < n; ++i) {
            count[s1[i] - 'a']++;
            count[s2[i] - 'a']--;
        }
        for (int i = 0; i < 26; ++i) {
            if (count[i] != 0) return false;
        }

        for (int len = 1; len < n; ++len) {
            if (isScramble(s1.substr(0, len), s2.substr(0, len)) && 
                isScramble(s1.substr(len), s2.substr(len))) {
                return true;
            }
            if (isScramble(s1.substr(0, len), s2.substr(n - len)) && 
                isScramble(s1.substr(len), s2.substr(0, n - len))) {
                return true;
            }
        }
        return false;
    }
};

File: MATTHEW/leetcode/AIcoding/4.GPT-4o/java_solutions/87-scramble-string.java
class Solution {
    public boolean isScramble(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        if (s1.equals(s2)) return true;

        int[] count = new int[26];
        for (char c : s1.toCharArray()) count[c - 'a']++;
        for (char c : s2.toCharArray()) {
            if (--count[c - 'a'] < 0) return false;
        }

        int n = s1.length();
        for (int len = 1; len < n; len++) {
            if (isScramble(s1.substring(0, len), s2.substring(0, len)) && 
                isScramble(s1.substring(len), s2.substring(len))) {
                return true;
            }
            if (isScramble(s1.substring(0, len), s2.substring(n - len)) && 
                isScramble(s1.substring(len), s2.substring(0, n - len))) {
                return true;
            }
        }

        return false;
    }
}